content/blog/2020/2020-10-15-document-embedders.md

```diff
@@ -0,0 +1,73 @@
+Embedding models are mostly based on neural networks.
+
+Document embedding is computed in two steps. First, each word is embedded with the word embedding then word embeddings are aggregated. The most common type of aggregation is the average over each dimension.
+
+## Why and when should we use embedders?
+
+Compared to [bag-of-words](/widget-catalog/text-mining/bagofwords-widget), which counts the number of appearances of each token in the document, embeddings have two main advantages:
+
+- **They do not have a dimensionality problem** The result of bag-of-words is a table which has the number of features equal to the number of unique tokens in all documents in a corpus. Large corpora with long texts result in a large number of unique tokens. It results in huge tables which can exceed memory in the computer. Huge tables also increase the learning and evaluation time of machine learning models. Embedders have constant dimensionality of the vector, which is 300 for fastText embeddings that Orange uses. 
+- **Most of the preprocessing is not required** In case of bag-of-words approach, we solve the dimensionality problem with the text preprocessing where we remove tokens (e.g. words) that seem to be less important for the analysis. It can also cause the removal of some important tokens. When using embedders, we do not need to remove tokens, so we are not losing the accuracy this way. Also most of the basic preprocessing can be omitted (such as normalization) in case of fastText embeddings.
+- **They can be pretrained** Word embedding models can be pretrained on large corpora with billions of tokens. That way, they capture the significant characteristics of the language and produce the embeddings of high quality. Pretrained models are then used to obtain embeddings of smaller datasets. Our Document Embedding widget uses pretrained fastText models and is suitable for corpora of any size.
+
+The shortcoming of the embedders is that they are difficult to understand. For example, when we use a bag-of-words, we can easily observe which tokens are important for classification with [Nomogram widget](/widget-catalog/visualize/nomogram/) since tokens themselves are features. In the case of document embeddings, features are numbers which are not understandable to human by themselves.
+
+## Document Embedding widget
+
+Orange now offers document embedders through Document Embedding widget. We decided to use [fastText pretrained embedders](https://fasttext.cc/docs/en/crawl-vectors.html), which support 157 languages. Orange's Document Embedding widget currently supports 31 most common languages.
+
+{{% workflow-screenshot src="/blog_img/2020/2020-10-15-document-embedding-widget.png" %}}
+
+In the widget, the user sets the language of documents and the aggregation method -- it is how embeddings aggregate embeddings for each word in document embedding. 
```

Comment:
how embeddings aggregate embeddings? :O

Reply:
That was bad. Hope it is more understandable now. If you have any better idea please let me know.